windows/src/developer/samples/imsample/imsample.cpp

```diff
@@ -566,6 +587,9 @@ extern "C" BOOL __declspec(dllexport) WINAPI DFText(HWND hwndFocus, WORD KeyStro
 {
 	WCHAR buf[32];
 	if(!curkbd) return FALSE;
+  if (!curkbd->groups) {
+    return FALSE;
+  }
```

Comment:
```suggestion
	if (!curkbd || !curkbd->groups) {
    return FALSE;
  }
```

Suggest doing this throughout your changes -- no need to split these failure conditions up.